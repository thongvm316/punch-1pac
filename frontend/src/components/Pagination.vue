<template>
  <ul class="pagination mt-4">
    <li
      v-show="pager.current_page > 1"
      ref="pagePrev"
      class="page-item"
    >
      <a @click="go(pager.current_page - 1)"><svg
        xmlns="http://www.w3.org/2000/svg"
        fill="currentColor"
        viewBox="0 0 24 24"
      ><path d="M16.67 0l2.83 2.829-9.339 9.175 9.339 9.167-2.83 2.829-12.17-11.996z" /></svg></a>
    </li>
    <li
      v-for="(n, key) in items"
      ref="pageNumber"
      :key="key"
      class="page-item"
      :class="{ active: n === pager.current_page }"
    >
      <a
        href="#"
        @click.prevent="go(n)"
      >{{ n }}</a>
    </li>
    <li
      v-show="pager.current_page + 1 <= pager.total_pages"
      ref="pageNext"
      class="page-item"
    >
      <a @click="go(pager.current_page + 1)"><svg
        xmlns="http://www.w3.org/2000/svg"
        fill="currentColor"
        viewBox="0 0 24 24"
      ><path d="M7.33 24l-2.83-2.829 9.339-9.175-9.339-9.167 2.83-2.829 12.17 11.996z" /></svg></a>
    </li>
  </ul>
</template>

<script>
export default {
  name: 'Pagination',

  props: {
    action: {
      type: String,
      required: true
    },
    namespace: {
      type: String,
      required: true
    },
    params: {
      type: Object,
      required: false
    }
  },

  data() {
    return {
      step: 2,
      items: []
    }
  },

  computed: {
    pager() {
      return this.$store.state[this.namespace].pager
    }
  },

  watch: {
    pager: {
      handler: function() {
        this.items = this.buildItems()
      },
      deep: true
    }
  },

  mounted() {
    this.items = this.buildItems()
  },

  methods: {
    go(n) {
      document.activeElement.blur()
      this.fetchItems(n)
    },

    buildItems() {
      let items = []
      let first = 1
      let last = this.pager.total_pages
      if (this.pager.current_page > this.step) {
        first = this.pager.current_page - this.step
      }
      last = this.pager.current_page + this.step <= this.pager.total_pages ? this.pager.current_page + this.step : this.pager.total_pages
      for (let i = first; i <= last; i++) {
        items.push(i)
      }
      return items
    },

    fetchItems(page) {
      this.$store.dispatch(`${this.namespace}/${this.action}`, { page, per_page: this.pager.per_page, ...this.params })
    }
  }
}
</script>
