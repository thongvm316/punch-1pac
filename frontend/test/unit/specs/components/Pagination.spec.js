import { shallowMount } from '@vue/test-utils'
import wrapperOps from '../../supports/wrapper'
import setComputed from '../../supports/set-computed'
import Pagination from '@/components/Pagination'

const go = jest.spyOn(Pagination.methods, 'go')
const buildItems = jest.fn().mockReturnValue([1, 2, 3])
const fetchItems = jest.spyOn(Pagination.methods, 'fetchItems')
const pager = {
  current_page: 2,
  next_page: 3,
  prev_page: 1,
  per_page: 20,
  total_pages: 3
}

const localWrapperOps = {
  ...wrapperOps,
  propsData: {
    namespace: 'groupRequests',
    action: 'getRequests'
  },

  computed: {
    pager() {
      return {
        ...pager,
        current_page: 1
      }
    }
  },

  stubs: {
    PIcoPrevArrow: true,
    PIcoNextArrow: true
  },

  methods: {
    buildItems
  }
}

describe('Pagination.vue', () => {
  let wrapper, prevPageButton, nextPageButton

  beforeEach(() => {
    wrapper = shallowMount(Pagination, localWrapperOps)
  })

  afterEach(() => { wrapper.vm.$destroy() })

  describe('when Pagination was mounted', () => {
    it('should render correctly', () => {
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper.isVueInstance()).toBeTruthy()
      expect(buildItems).toHaveBeenCalled()
      expect(wrapper).toMatchSnapshot()
    })

    it('page 2 should be actived', async () => {
      setComputed(wrapper, { pager })
      await wrapper.vm.$nextTick()

      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('when click on page item', () => {
    it('should call go method when click page 2', async () => {
      wrapper.findAll({ ref: 'pageNumber' }).at(1).find('a').trigger('click')
      await wrapper.vm.$nextTick()

      expect(go).toHaveBeenCalledWith(2)
      expect(fetchItems).toHaveBeenCalledWith(2)
    })
  })

  describe('when current page is last', () => {
    it('should not show button next page', async () => {
      setComputed(wrapper, {
        pager: {
          ...pager,
          current_page: 3
        }
      })
      await wrapper.vm.$nextTick()

      expect(wrapper).toMatchSnapshot()
    })
  })
})
