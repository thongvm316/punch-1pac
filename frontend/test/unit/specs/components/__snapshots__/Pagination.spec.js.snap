// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination.vue when Pagination was mounted page 2 should be actived 1`] = `
<ul class="pagination mt-4">
  <li class="page-item" style=""><a>
      <picoprevarrow-stub></picoprevarrow-stub>
    </a></li>
  <li class="page-item"><a href="#">1</a></li>
  <li class="page-item active"><a href="#">2</a></li>
  <li class="page-item"><a href="#">3</a></li>
  <li class="page-item"><a>
      <piconextarrow-stub></piconextarrow-stub>
    </a></li>
</ul>
`;

exports[`Pagination.vue when Pagination was mounted should render correctly 1`] = `
<ul class="pagination mt-4">
  <li class="page-item" style="display: none;"><a>
      <picoprevarrow-stub></picoprevarrow-stub>
    </a></li>
  <li class="page-item active"><a href="#">1</a></li>
  <li class="page-item"><a href="#">2</a></li>
  <li class="page-item"><a href="#">3</a></li>
  <li class="page-item"><a>
      <piconextarrow-stub></piconextarrow-stub>
    </a></li>
</ul>
`;

exports[`Pagination.vue when current page is last should not show button next page 1`] = `
<ul class="pagination mt-4">
  <li class="page-item" style=""><a>
      <picoprevarrow-stub></picoprevarrow-stub>
    </a></li>
  <li class="page-item"><a href="#">1</a></li>
  <li class="page-item"><a href="#">2</a></li>
  <li class="page-item active"><a href="#">3</a></li>
  <li class="page-item" style="display: none;"><a>
      <piconextarrow-stub></piconextarrow-stub>
    </a></li>
</ul>
`;
